### DISCUSSION TITLE
Why doesn't this code compile?

### DISCUSSION THREAD

**Discussion by obiwanjacobi** [dotnet/roslyn#65354] [Q&A]
I recently found myself typing something like the following, completely confident this would be correct.
Apparently it is not.

```csharp
public class C {
    public virtual void Method<T>(T? p)
        where T : class
    { }
}

public class D : C {
    public override void Method<T>(T? p)
    {
        base.Method(p);
    }
}
```

I get [CS0115](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0115) and [CS0452](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0452) errors.
Running VS 2022 17.4.0

My question is: why?

**Comment by CyrusNajmabadi** [comment]
You need to write:

```c#
public class C {
    public virtual void Method<T>(T? p)
        where T : class
    { }
}

public class D : C {
    public override void Method<T>(T? p) where T : class
    {
        base.Method(p);
    }
}
```

At teh override point, without the constraints, the compiler cannot determine what method you are overriding.  It sees `<T>(T? p)` and goes "that could be an unconstrained T, with a defaultable parameter.  Or it could be a class T, with a nullable parameter. Or it could be a struct T, with a `Nullable<T>` parameter".  Since it doesn't know, it doesn't know what it is overriding. 

The fix is to put the constraint back in so it's clear which interpretation the override has.  WHich then allows the compiler to figure out what method is being overridden.

**Reply by obiwanjacobi** [reply]
Thanks. I always thought that overrides were resolved by looking at candidates and then picking the (best) one that fits...?
Anyway my example left out a crucial detail - below the new code. But I think I understand the mechanics now. I also think it looks wrong ;-)

**Reply by ufcpp** [reply]
```cs
public class C
{
    public virtual void Method<T>(T? p)
        where T : class
    { }

    public virtual void Method<T>(T? p)
        where T : struct
    { }
}
```

**Comment by obiwanjacobi** [comment]
Turns out the simplified example I've tried to create from my real situation left a a crucial aspect. My initial sample can be made to work by repeating the type constraint at the override. You can do that if the constraint is `class` (or `struct`). But not when it is another class (as it was in my actual situation).

```csharp
public class X {}

public class C {
    public virtual void Method<T>(T? p)
        where T : X
    { }
}

public class D : C {
    public override void Method<T>(T? p)
        where T : X   // <= CS0460
    {
        base.Method(p);
    }
}
```

The only way to get this to compile is to remove the nullable `?` from the parameter (and not repeating the type constraint).

```csharp
public class X {}

public class C {
    public virtual void Method<T>(T? p)
        where T : X
    { }
}

public class D : C {
    public override void Method<T>(T p)
    {
        base.Method(p);
    }
}
```

Is that because the Nullable Reference stuff is just attributes and doesn't count as a (different) method signature in this case?

**Reply by svick** [reply]
It's definitely not intuitive, but [it seems specifying `where T : class` in the override works even in this case](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABAzAAmwCZ8ANfAbwF8BYAKHr0JIGFL79PCCwYpgArgEMANoQAs+ALIRgACwD2KADwAVAHwAKVQH58ABwCUHLqYDuc6BHyr8IMic4V8tOq8YFi+ACJ38bCkdufAVIKFgUa2xJGXklNS1dA2M6U3NLKGtbe2IgwNS0zgAjIQBnCAA6WMUUTSMAbiDXKiA===):

```c#
public class X {}

public class C {
    public virtual void Method<T>(T? p)
        where T : X
    { }
}

public class D : C {
    public override void Method<T>(T? p)
        where T : class
    {
        base.Method(p);
    }
}
```

**Reply by obiwanjacobi** [reply]
Haha, I would have never thought of that because I would have never thought that would be legal. Thanks!
